pandas/core/arrays/interval.py

```diff
@@ -0,0 +1,856 @@
+
+        # Need to ensure that left and right are updated atomically, so we're
+        # forced to copy, update the copy, and swap in the new values.
+        left = self.left.copy(deep=True)
+        right = self.right.copy(deep=True)
+
+        left.values[key] = value.left
+        right.values[key] = value.right
+
+        self._left = left
+        self._right = right
+
+    def fillna(self, value=None, method=None, limit=None):
+        if method is not None:
+            raise TypeError('Filling by method is not supported for '
+                            'IntervalArray.')
+        if limit is not None:
+            raise TypeError('limit is not supported for IntervalArray.')
+
+        if not isinstance(value, Interval):
+            msg = ("Interval.fillna only supports filling with scalar "
+                   "'value'. Got a '{}' instead of a 'pandas.Interval'."
+                   .format(type(value).__name__))
+            raise TypeError(msg)
+
+        value = getattr(value, '_values', value)
+
+        left = self.left.fillna(value=value.left)
+        right = self.right.fillna(value=value.right)
+        return self._shallow_copy(left, right)
+
+    @property
+    def dtype(self):
+        return IntervalDtype.construct_from_string(str(self.left.dtype))
+
+    def astype(self, dtype, copy=True):
+        dtype = pandas_dtype(dtype)
+        if is_interval_dtype(dtype) and dtype != self.dtype:
+            try:
+                new_left = self.left.astype(dtype.subtype)
+                new_right = self.right.astype(dtype.subtype)
+            except TypeError:
+                msg = ('Cannot convert {dtype} to {new_dtype}; subtypes are '
+                       'incompatible')
+                raise TypeError(msg.format(dtype=self.dtype, new_dtype=dtype))
+            return self._shallow_copy(new_left, new_right)
+        elif is_interval_dtype(dtype):
+            if copy:
+                return self.copy()
+            else:
+                return self
+        elif is_categorical_dtype(dtype):
+            return Categorical(self.values)
+        # TODO: This try/except will be repeated.
+        try:
+            return self.values.astype(dtype, copy=copy)
+        except (TypeError, ValueError):
+            msg = 'Cannot cast {name} to dtype {dtype}'
+            raise TypeError(msg.format(name=type(self).__name__, dtype=dtype))
+
+    @classmethod
+    def _concat_same_type(cls, to_concat):
+        closed = set(interval.closed for interval in to_concat)
+        if len(closed) != 1:
+            raise ValueError("Intervals must all be closed on the same side.")
+        closed = closed.pop()
+
+        # TODO: avoid intermediate list
+        left = np.concatenate([interval.left for interval in to_concat])
+        right = np.concatenate([interval.right for interval in to_concat])
+        return cls._simple_new(left, right, closed=closed, copy=False)
+
+    def _shallow_copy(self, left=None, right=None):
```

Comment:
is this a method in EA? it should be

Reply:
As long as this is only uses internally within the IntervalArray class, it is not necessarily needed to make this part of the official ExtensionArray interface. 

What are the advantages in doing so? (also the signature will be different for each array type)